rotkehlchen/chain/evm/decoding/rainbow/decoder.py

```diff
@@ -0,0 +1,380 @@
+        fee_raw = int.from_bytes(unwrap_log.data) - in_amount
+
+        # Create the event for the fee
+        fee_event = self._create_fee_event(
+            tx_hash=transaction.tx_hash,
+            timestamp=transaction.timestamp,
+            fee_raw=fee_raw,
+            fee_asset=native_token,
+            sender=sender,
+        )
+        decoded_events.append(fee_event)
+
+        return
+
+    def _maybe_decode_swap_token_to_token(
+            self,
+            transaction: EvmTransaction,
+            decoded_events: list['EvmEvent'],
+            all_logs: list[EvmTxReceiptLog],
+            sender: ChecksumEvmAddress,
+            out_event: 'EvmEvent',
+    ) -> None:
+        """Decodes a swap transaction from token to native token.
+        Via the first and last transfers logs
+        """
+        try:
+            token_address = out_event.asset.resolve_to_evm_token().evm_address
+        except WrongAssetType:
+            token_address = None
+
+        if token_address is None:
+            log.error(
+                f'Failed to decode swap from token to native token '
+                f'for tranction {transaction.tx_hash.hex()} due the or token_address is null.',
+            )
+            return
+
+        sell_amount = swap_amount = None
+        for tx_log in all_logs:
+            if (
+                tx_log.topics[0] == ERC20_OR_ERC721_TRANSFER and
+                tx_log.address == token_address
+            ):
+                # Find the transfer log from the user wallet to router
+                if bytes_to_address(tx_log.topics[2]) == self.router_address:
+                    sell_amount = int.from_bytes(tx_log.data)
+
+                # Find the transfer log from the router to the swap provider
+                elif bytes_to_address(tx_log.topics[1]) == self.router_address:
+                    swap_amount = int.from_bytes(tx_log.data)
+
+                if swap_amount and sell_amount:
+                    break
+
+        if not (sell_amount and swap_amount):
+            log.error(
+                f'Failed to decode swap from token to token '
+                f'for tranction {transaction.tx_hash.hex()} due the sell_amount '
+                'or swap_amount being null.',
+            )
+            return
+
+        fee_event = self._create_fee_event(
+            tx_hash=transaction.tx_hash,
+            timestamp=transaction.timestamp,
+            fee_raw=sell_amount - swap_amount,
+            fee_asset=out_event.asset.resolve_to_evm_token(),
+            sender=sender,
+        )
+        decoded_events.append(fee_event)
+
+        return
+
+    def _process_swap(self,
+            transaction: 'EvmTransaction',
+            decoded_events: list['EvmEvent'],
+            all_logs: list['EvmTxReceiptLog'],
+    ) -> list['EvmEvent']:
+        sender = transaction.from_address
+        in_event, out_event = self._get_in_out_events(decoded_events, sender)
```

Comment:
explain the docstring what you do here and what you are looking in the logs for the wrapping so the reader can follow the code easily